### PR TITLE
Scripting API cleanup based upon review feedback

### PR DESCRIPTION
**Pull request by tmat**
Implements most of the API review feedback.

Remaining work:
- Compilation API cleanup
- Script.Build return diagnostics instead of throwing

### CODE REVIEW

**Comment by tmat** [comment]
@mattwar @jaredpar @Pilchie @DustinCampbell @ManishJayaswal

File: src/Scripting/Core/ScriptOptions.cs [hunk 1]
```diff
@@ -182,193 +183,50 @@ private ScriptOptions(ScriptOptions other)
...
-        /// </remarks>
-        public ScriptOptions WithDefaultSourceResolution(params string[] searchPaths) =>
-           WithDefaultSourceResolution(searchPaths.AsImmutableOrEmpty());
-
-        /// <summary>
         /// Creates a new <see cref="ScriptOptions"/> with specified <see cref="SourceResolver"/>.
         /// </summary>
-        public ScriptOptions WithCustomSourceResolution(SourceReferenceResolver resolver) =>
+        public ScriptOptions WithSourceResolver(SourceReferenceResolver resolver) =>
             SourceResolver == resolver ? this : new ScriptOptions(this) { SourceResolver = resolver };
```
Comment (KevinH-MS) [inline src/Scripting/Core/ScriptOptions.cs]: Now that I'm looking at this, it doesn't feel quite right.  I think this should either throw or null out the SourceResolver.  Why would we no-op on null?

Comment (KevinH-MS) [inline src/Scripting/Core/ScriptOptions.cs]: Sorry, I didn't read that right.

File: src/Scripting/Core/ScriptSourceResolver.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.Scripting
+{
+    using static ParameterValidationHelpers;
+
+    public sealed class ScriptSourceResolver : SourceFileResolver, IEquatable<ScriptSourceResolver>
```
Comment (KevinH-MS) [inline src/Scripting/Core/ScriptSourceResolver.cs]: I haven't bothered to check the references, but does this need to be public?  It feels like an implementation detail that people could trivially duplicate if they're wanting to implement their own host.

Comment (tmat) [inline src/Scripting/Core/ScriptSourceResolver.cs]: This is added for discoverability and consistency with ScriptReferenceResolver. Scripting users should only need to import Microsoft.CodeAnalysis.Scripting(.CSharp);

File: src/Scripting/Core/ScriptState.cs [hunk 3]
```diff
@@ -91,7 +91,7 @@ private ImmutableArray<ScriptVariable> CreateVariables()
                 foreach (var field in state.GetType().GetTypeInfo().DeclaredFields)
                 {
                     // TODO: synthesized fields of submissions shouldn't be public
-                    if (field.IsPublic && field.Name.Length > 0 && char.IsLetterOrDigit(field.Name[0]))
+                    if (field.IsPublic && field.Name.Length > 0 && (char.IsLetterOrDigit(field.Name[0]) || field.Name[0] == '_'))
```
Comment (KevinH-MS) [inline src/Scripting/Core/ScriptState.cs]: Consider test for "_blah" case (make sure it works end-to-end)

**Comment by KevinH-MS** [comment]
:+1:

**Comment by tmat** [comment]
test eta please

**Comment by mattwar** [comment]
:+1:

**Comment by tmat** [comment]
test this please

**Pull request merged**